Search for keys in a directory tree that matches their scope

Saford91 opened the pull request:
For example, the last attempt to find key foo::bar::baz in a directory
common.d will search for common.d/foo/bar/baz

Saford91 committed:
Search for keys in a hierarchy that matches their scope

For example, the last attempt to find key foo::bar::baz in a directory
common.d will search for common.d/foo/bar/baz

Saford91 committed:
Use gsub instead of sub. Remove debug line.

Saford91 committed:
Account for the target key being a directory instead of a file

ffrank commented:
Sound idea! I feel that splitting by `::` and then feeding the components to `File.join` would be more appropriate than `gsub`. What do you think?

Do you feel up to adding a test or two for this functionality?

The spec failures look sad, but I don't think they are related to your change.

Saford91 commented:
I'll add some tests.

Using split instead of gsub seems inconsequential. I'm not opposed to it, but is there a practical difference?

ffrank commented:
`File.join` is platform agnostic. Your implementation assumes that `/` is an acceptable path separator (I'm not even sure this is the case in Windows, e.g. ...hiera runs on Windows right? 😇 )

Saford91 committed:
Maintain platform agnosticism

Using split instead of gsub giuarantees the path will always have a path
separator appropriate for the operating system

ffrank commented:
Thanks for indulging me :)
I think there's a problem now. The component array is getting passed to `validate_key_lookup!`. Will this work? I suppose this needs to be updated as well. Cheers!

Saford91 commented:
Hey @ffrank , I think the `validate_key_lookup!` will still work as intended. If `key` is an array, `File.join` will join all the array members together. I don't see a need to update it, do you?
